Assignment 0 (#3)
LaunchCodeEducation/csharp-web-dev-curriculum: content/assignments/hello-world/auto-grading-test.md

   You will not need to open the `Using.cs` file.  This file connects the testing project to the testing library.  This is not a test class.
   {{% /notice %}}

   A **test** is a single method in a test file with `[TestMethod]` above it. 

      ```csharp
      [TestClass]
      public class AutogradingTests
      {
         [TestMethod]
         public void TestToDemonstrateWhatATestMethodLooksLike()
         {
            //test code here
         }
      }
      ```

      A test class may contain many tests. 

      ```csharp
      [TestClass]
      public class AutogradingTests
      {
         [TestMethod]
         public void TestToDemonstrateWhatATestMethodLooksLike()
         {
            //test code here
         }

         [TestMethod]
         public void SecondTestToDemonstrateWhatATestMethodLooksLike()
         {
            //test code here
         }
      }
   ```
1. Run all of your tests using the Test Explorer window.
   1. Mac Users: 
      1. **View** > **Tests**  
      1. Select the green triangle on the left of the Test Explorer Window. 
      1. In the Test Explorer, step through the file tree until you find the failed test. 

>       1. If you need more help running the tests, checkout this [guide](https://learn.microsoft.com/en-us/visualstudio/mac/testing?view=vsmac-2022#running-tests) for running tests in Visual Studio for Mac

Reviewer: ```suggestion
      1. If you need more help running the tests, checkout this [guide](https://learn.microsoft.com/en-us/visualstudio/mac/testing?view=vsmac-2022#running-tests) for running tests in Visual Studio for Mac.
```

   1. Windows Users 
      1. **Test** > **Windows** > **Test Explorer**
      1. Select the `Run All` icon to run all of your tests. 
      1. In the Test Explorer, step through your file tree until you find the failed test.
      1. If you need more help running the tests, checkout this [guide](https://learn.microsoft.com/en-us/visualstudio/test/run-unit-tests-with-test-explorer?view=vs-2022) for running tests in Visual Studio.

1. Visual Studio opens a Test Results Pane every time it runs a test. This pane provides you with detailed information about which tests passed and which tests failed.  The pane will also provide you with output for any failed tests, including what was expected and what was actually received.  This is a great place to look when dealing with a failing test.

>    Now that you have run the tests, you should see that one of the tests has failed and the other passed. We have not touched any code yet. Use the Test Results Pane to navigate into which specific test has failed.

Reviewer: ```suggestion
   Now that you have run the tests, you should see that one of the tests has failed and the other passed. We have not touched any code yet. Using the Test Results Pane double click on the test name to navigate to the test code that has failed.
```

Reviewer: This sentence just seems off.  What are we suppose to be looking at?  There is a little carrot I can click that expands and gives me information on why the test failed but I also get that to the right of the tests as well.

Author: in this example, we only have 1 test class with 2 tests. in A2 and A4 you may have up to 8 tests in a test class.  those assignments also have multiple test classes.  there is a lot more scrolling to be done.  while it might seem obvious, i like the word "navigate" because the tests and test classes will get more complex as the assignments grow.

>    In this case, the `TestIfSayHelloReturnsCorrectValue()` test, expected our code to output `Hello, World!` Instead, the method is returning is `&&&` causing the test to fail. Be sure to explore your test output. The output will show you what was expected and what was actually received.  Often messages will be provided for a failed test. In this test, the message is "Incorrect message displayed".

Reviewer: ```suggestion
   In this case, the `TestIfSayHelloReturnsCorrectValue()` test, expected our code to output `Hello, World!` Instead, the method is returning `&&&` causing the test to fail. Be sure to explore your test output. The output will show you what was expected and what was actually received.  Often messages will be provided for a failed test. In this test, the message is "Incorrect message displayed".
```

   The `TestSayHelloReturnTypeIsString()` test passed. This should pass because the `SayHello()` method is a string method. You can verify this in line 8 of the `SayHelloClass.cs`. Use this test to help you explore the Test Results Pane and testing process. 

1. Try running each test individually before moving on to the next step. You can select individual tests by right-clicking on the test name either directly in the test class or the Test Explorer window.

   You will still have the same results as before, but now you have the skills to run tests all together or one at a time. This is something that may come in handy as you work on larger assignments that may have many tests.

> ### Next Steps

Reviewer: I'm not a fan of the Next Steps at the bottom of each page.  Could just be a sentence to move to the next page.
